### PR TITLE
Properly support PEP 695 generics syntax

### PR DESCRIPTION
**Pull request by Viicos**
Best reviewed commit per commit.

Reword relevant docstrings and comments.

Fixes https://github.com/pydantic/pydantic/issues/11031.
Fixes https://github.com/pydantic/pydantic/issues/9782.

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: docs/concepts/models.md
Comment (Viicos) [inline docs/concepts/models.md:1]: Best to be reviewed visually. The example was simplified a bit (avoid using lists, etc)

File: pydantic/_internal/_namespace_utils.py [hunk 1]
```diff
@@ -125,11 +125,7 @@ def ns_for_function(obj: Callable[..., Any], parent_namespace: MappingNamespace
     # passed as a separate argument. However, internally, `_eval_type` calls
     # `ForwardRef._evaluate` which will merge type params with the localns,
     # essentially mimicking what we do here.
-    type_params: tuple[_TypeVarLike, ...]
-    if hasattr(obj, '__type_params__'):
-        type_params = obj.__type_params__
-    else:
-        type_params = ()
+    type_params: tuple[_TypeVarLike, ...] = getattr(obj, '__type_params__', ())
```
Comment (sydney-runkle) [inline pydantic/_internal/_namespace_utils.py:128]: 👍

File: pydantic/_internal/_typing_extra.py [hunk 2]
```diff
@@ -524,35 +524,62 @@ def is_generic_alias(tp: Any, /) -> bool:
...
+
+    # note, we don't copy frame.f_locals here (or during the last return call), because we don't expect the namespace to be
+    # modified down the line if this becomes a problem, we could implement some sort of frozen mapping structure to enforce this.
     if force:
         return frame.f_locals
 
-    # if either of the following conditions are true, the class is defined at the top module level
-    # to better understand why we need both of these checks, see
-    # https://github.com/pydantic/pydantic/pull/10113#discussion_r1714981531
+    # if either of the following conditions are true, the class is defined at the top module level.
```
Comment (sydney-runkle) [inline pydantic/_internal/_typing_extra.py:580]:
```suggestion
    # If either of the following conditions are true, the class is defined at the top module level.
```

File: tests/test_forward_ref.py [hunk 3]
```diff
@@ -1365,3 +1365,78 @@ class Model(BaseModel, A):
         ReturnedModel = inner()  # noqa: F841
 
     assert module_1.ReturnedModel.__pydantic_complete__ is False
+
+
+# Tests related to forward annotations evaluation coupled with PEP 695 generic syntax:
```
Comment (sydney-runkle) [inline tests/test_forward_ref.py:1370]: Is this needed, given the skipifs below?

File: tests/test_typing.py [hunk 4]
```diff
@@ -160,13 +146,17 @@ def test_func_ns_excludes_default_globals() -> None:
         assert default_global_var not in func_ns
 
 
-module_foo = 'global_foo'
-module_ns = parent_frame_namespace(parent_depth=1)
+def test_module_ns_is_none(create_module) -> None:
```
Comment (sydney-runkle) [inline tests/test_typing.py:149]: Can we add a test with `force=True`, and guarantee that the result is not `None` as well?

File: docs/concepts/models.md [hunk 5]
```diff
@@ -648,62 +648,96 @@ Here are some additional notes on the behavior of [`model_construct()`][pydantic
...
-    Response[int](data='value')
-except ValidationError as e:
-    print(e)
-    """
-    1 validation error for Response[int]
-    data
-      Input should be a valid integer, unable to parse string as an integer [type=int_parsing, input_value='value', input_type=str]
-    """
-```
+=== "Python 3.12 and above (new syntax)"
```
Comment (sydney-runkle) [inline docs/concepts/models.md:704]:
```suggestion
=== "Python 3.12+ (new type parameter syntax)"
```

**Review by sydney-runkle — APPROVED** [review]
New docs look great. I left a few minor requests for changes.

I'm impressed that the diff is this minimal for supporting this syntax!

File: tests/test_forward_ref.py (hunk 3 above)
Comment (sydney-runkle) [reply tests/test_forward_ref.py:1370]: Fine if you want to keep it, just wanted to check :)

Comment (Viicos) [reply tests/test_forward_ref.py:1370]: The comment acts as some kind of "grouping" of tests (which in reality is a sign that we should split our test files :sweat_smile:), and at some point (in a really long time) the skipifs will be removed but the comment should remains

File: docs/concepts/models.md (hunk 5 above)
Comment (Viicos) [reply docs/concepts/models.md:704]: I used the same wording as our upgrade plugin.

Otherwise, changed to _new type parameter syntax_
